### PR TITLE
Flying robots, projectiles, and food

### PR DESCRIPTION
**Pull request by NathanHinthorne**
### Additions
- Added overseer bot
- Added oculi bot
- Added sound effects for snake attacks, as well as projectiles
- Implemented missile (blows up), laser, and sonic wave (pushes Chad back)
- Added new music
- Made death animations for robots
- Added health counter
- Added roast turkey (a type of FoodDrop)
- Improved endzone by adding new bots to it, as well as changing its layout a bit.

### Fixes
- Implemented most suggested changes from my last PR
-  Balanced game by tweaking enemy stats, adding new drops, and making a chance for enemy projectiles to misfire a bit to the right or left

### CODE REVIEW

File: js/classes/entities/midground/chad.js [hunk 1]
```diff
@@ -242,25 +244,59 @@ class Chad {
         }
     };
 
+    knockback(direction, amount) {
```
Comment (b1gd3vd0g) [inline js/classes/entities/midground/chad.js:247]: oh this is cool, I'm excited to see how it looks in the game

File: js/classes/entities/midground/drops/ammo_drop.js [hunk 2]
```diff
@@ -10,7 +10,7 @@ class AmmoDrop {
      * @param {Vector} pos The position at which the AmmoDrop should start.
      * @param {number} type The type of AmmoDrop that should be generated. AmmoDrop.ROCK, .BOMB, etc.
      */
-    constructor(pos, type, amount = 1, hasGravity = true, popInAir = false) {
+    constructor(pos, type, amount = 1, hasGravity = true, popInAir = true) {
```
Comment (b1gd3vd0g) [inline js/classes/entities/midground/drops/ammo_drop.js:13]: what does popInAir mean?

**Review by b1gd3vd0g — APPROVED** [review]
Real nice work you got here Nathan! I'm excited to check out these new features.

File: js/classes/entities/foreground/food_item.js [hunk 3]
```diff
@@ -85,6 +85,10 @@ class FoodItem {
     static get GIANT_MUSHROOM() {
         return 7;
     }
+
+    static get ROAST_TURKEY() {
+        return 8;
```
Comment (calebkrauter) [inline js/classes/entities/foreground/food_item.js:90]: Woah a new food!

File: js/classes/entities/foreground/ui/death_screen.js [hunk 4]
```diff
@@ -60,7 +60,7 @@ class DeathScreen {
      * Respond to the respawn button click.
      */
     handleRespawnButtonClick() {
-        CHAD.health = Chad.MAX_HEALTH;
+        CHAD.health = CHAD.maxHealth;
```
Comment (calebkrauter) [inline js/classes/entities/foreground/ui/death_screen.js:63]: What's up with this change? Does his max health increase now?

File: js/classes/entities/foreground/ui/hud.js [hunk 5]
```diff
@@ -597,20 +598,25 @@ class HudHealthBar {
...
         // draw background
         CTX.fillStyle = "#000000";
-        CTX.fillRect(this.pos.x, this.pos.y, HudHealthBar.SIZE.x, HudHealthBar.BAR_HEIGHT);
+        CTX.fillRect(this.pos.x, this.pos.y, this.length, HudHealthBar.BAR_HEIGHT);
 
         // draw the actual health bar
         CTX.fillStyle = "#ff0000";
         CTX.fillRect(this.pos.x + HudHealthBar.PADDING, this.pos.y + HudHealthBar.PADDING,
-            (HudHealthBar.SIZE.x - HudHealthBar.PADDING * 2) * CHAD.health / Chad.MAX_HEALTH,
+            (this.length - HudHealthBar.PADDING * 2) * CHAD.health / CHAD.maxHealth,
```
Comment (calebkrauter) [inline js/classes/entities/foreground/ui/hud.js:613]: Why did you change HudHealthBar to 'this'? Seems a little less clear.

File: js/classes/entities/midground/chad.js [hunk 6]
```diff
@@ -242,25 +244,59 @@ class Chad {
...
+        this.isKnockedBack = true;
+    }
+
+    manageKnockback() {
+        if (this.isKnockedBack) {
+            this.knockbackForce = Vector.multiply(this.knockbackForce, 0.5);
+            this.velocity = Vector.add(this.velocity, this.knockbackForce);
+
+            // If velocity is below a certain threshold, end knockback
+            if (Vector.magnitude(this.knockbackForce) < 0.1) {
```
Comment (calebkrauter) [inline js/classes/entities/midground/chad.js:260]: Nice!

File: js/classes/entities/midground/chad.js [hunk 7]
```diff
@@ -242,25 +244,59 @@ class Chad {
...
+        }
+    }
+
+
     /**
      * Increase the health of Chad by the provided amount
      * 
      * @param {number} amount the amount by which to increase Chad's health
      */
     restoreHealth(amount) {
```
Comment (calebkrauter) [inline js/classes/entities/midground/chad.js:272]: I like this. But will there be an issue with infinite health regen? He may become too powerful.

File: js/classes/entities/midground/chad.js [hunk 8]
```diff
@@ -318,7 +354,7 @@ class Chad {
             }
 
             // release wind particles every 0.05 seconds
```
Comment (calebkrauter) [inline js/classes/entities/midground/chad.js:356]: This comment is now outdated and if we update the magic numbers to clear variables it is unnecessary.

File: js/classes/entities/midground/drops/food_drop.js [hunk 9]
```diff
@@ -53,6 +53,13 @@ class FoodDrop {
         ASSET_MGR.playSFX(sfx.path, sfx.volume);
 
         switch (this.type) {
+            case FoodItem.ROAST_TURKEY:
```
Comment (calebkrauter) [inline js/classes/entities/midground/drops/food_drop.js:56]: Ah I see so max health only increases with roast turkey. Very good! You probably will want to make it hard to get multiple times since you can literally farm any zone right now for food. You may have already considered this. Are you making the big snake drop it?

File: js/classes/entities/midground/enemies/flying_enemy_base.js [hunk 10]
```diff
@@ -39,6 +39,17 @@ class FlyingEnemyBase {
...
 
+    /**
+     * Convert the relative path to an absolute path.
+     * 
+     * @param {Vector[]} path the relative path
+     * @param {Vector} pos the position of the enemy
+     * @returns {Vector[]} the absolute path
+     */
+    relativeToAbsolute(path, pos) {
+        return path.map((node) => Vector.add(node, pos));
```
Comment (calebkrauter) [inline js/classes/entities/midground/enemies/flying_enemy_base.js:50]: Can you explain to me how this works? I'm unfamiliar.

File: js/classes/entities/midground/enemies/oculi_bot.js [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+                    // if it's been long enough, start a new attack 
+                    this.animations[this.getFacing()]["attacking"].elapsedTime = 0;
+                    this.action = "attacking";
+                    this.lastAttack = Date.now() / 1000;
+                    this.dealtDamage = false;
+                } else if (this.action === "attacking" 
+                    && secondsSinceLastAttack > OculiBot.DAMAGE_DELAY && !this.dealtDamage) {
+                    // if we're at the proper point in our attack animation, deal damage
+    
+                    const misdirectVector = new Vector(Math.random() * 20 - 10, Math.random() * 20 - 10);
```
Comment (calebkrauter) [inline js/classes/entities/midground/enemies/oculi_bot.js:144]: What is a misdirectVector?

File: js/classes/entities/midground/enemies/overseer_bot.js [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+                    this.lastAttack = Date.now() / 1000;
+                    this.dealtDamage = false;
+                } else if (this.action === "attacking" 
+                    && secondsSinceLastAttack > OverseerBot.DAMAGE_DELAY && !this.dealtDamage) {
+                    // if we're at the proper point in our attack animation, deal damage
+    
+                    const misdirectVector = new Vector(Math.random() * 20 - 10, Math.random() * 20 - 10);
+                    const target = Vector.add(CHAD.getCenter(), misdirectVector);
+                    if (Math.random() < 0.5) {
+                        GAME.addEntity(ProjectileFactory.create(ProjectileFactory.MISSILE, this.pos, target));
```
Comment (calebkrauter) [inline js/classes/entities/midground/enemies/overseer_bot.js:147]: Can't wait to see this.

File: js/classes/entities/midground/weapon/bomb.js [hunk 13]
```diff
@@ -122,6 +123,11 @@ class Bomb {
         // do nothing?
     }
 
+    /** Called when the bomb collides with the player. */
+    onPlayerCollision(player) {
+        // do nothing?
```
Comment (calebkrauter) [inline js/classes/entities/midground/weapon/bomb.js:128]: What's up with this?

File: js/classes/entities/midground/weapon/broccoli.js [hunk 14]
```diff
@@ -17,7 +17,8 @@ class Broccoli {
             Broccoli.INITIAL_SPEED,
             Broccoli.WEIGHT,
             (block) => this.onBlockCollision(block),
-            (enemy) => this.onEnemyCollision(enemy)
+            (enemy) => this.onEnemyCollision(enemy),
```
Comment (calebkrauter) [inline js/classes/entities/midground/weapon/broccoli.js:20]: What is this syntax doing? Is this some sort of arrow function? I'm not super familiar with this in JS being in a parameter list.

File: js/classes/entities/midground/weapon/missile.js [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+            new Vector(0, Missile.SIZE.y * 6), 
+            Missile.SIZE, 2, 0.5); // 270 degree angle
+
+        this.animations["firing"]["rightUp"] = new Animator(
+            Missile.SPRITESHEET, 
+            new Vector(0, Missile.SIZE.y * 7), 
+            Missile.SIZE, 2, 0.5); // 315 degree angle
+    }
+
+    update() {
```
Comment (calebkrauter) [inline js/classes/entities/midground/weapon/missile.js:154]: Not a huge deal but don't we usually put the update method above the `loadAnimations()`? Might be good to stay consistent if that is correct.

File: js/classes/entities/midground/weapon/projectile_base.js [hunk 16]
```diff
@@ -62,11 +67,41 @@ class ProjectileBase {
...
         this.projectile.pos = Vector.add(this.projectile.pos, new Vector(pos.x, pos.y + this.projectile.yVelocity));
 
+        // update the rotation
+        let angle = Math.atan2(this.projectile.dir.y, this.projectile.dir.x);
+        if (angle < 0) {
+            angle += 2 * Math.PI; // convert to 0 to 2π range
+        }
+        this.projectile.rotation = angle;
+        
+        if (angle >= 0 && angle < Math.PI * 0.125 || angle >= Math.PI * 1.875) {
```
Comment (calebkrauter) [inline js/classes/entities/midground/weapon/projectile_base.js:77]: Wow this is a lot. If it works that's great but I can't help but wonder how this could be cleaned up. Perhaps a function with parameters which would be the values that change in the conditions such as 0.125 etc and whatever the increase or decrease in the value could be added to the value depending on the direction. For example the delta in the values you change from right to rightDown to Down, are they all the same delta value? Perhaps a pattern arises and this could be made into a generic function.

File: js/classes/entities/midground/weapon/projectile_factory.js [hunk 17]
```diff
@@ -7,7 +7,10 @@ class ProjectileFactory {
             [ProjectileFactory.SNOWBALL]: Snowball,
             [ProjectileFactory.SUS_SNOWBALL]: SusSnowball,
             [ProjectileFactory.BROCCOLI]: Broccoli,
-            [ProjectileFactory.WATER_BALLOON]: WaterBalloon
+            [ProjectileFactory.WATER_BALLOON]: WaterBalloon,
```
Comment (calebkrauter) [inline js/classes/entities/midground/weapon/projectile_factory.js:10]: Is this like some sort of map?

File: js/classes/entities/midground/weapon/slimeball.js [hunk 18]
```diff
@@ -106,6 +107,11 @@ class Slimeball {
         }
     }
 
+    /** Called when the slimeball collides with the player. */
```
Comment (calebkrauter) [inline js/classes/entities/midground/weapon/slimeball.js:110]: I keep seeing this.

File: js/classes/entities/midground/weapon/sword.js [hunk 19]
```diff
@@ -112,6 +112,12 @@ class Sword {
                     if (bb.collide(entity.boundingBox)) {
                         entity.takeDamage(this.damage * CHAD.damageMultiplier);
                         this.hasHit = true;
+
```
Comment (calebkrauter) [inline js/classes/entities/midground/weapon/sword.js:115]: Why is this commented out? Might be good to keep it in if it works until you find a better solution.

File: js/scripts/zone_loaders/village_zone_loaders.js [hunk 20]
```diff
@@ -141,6 +141,12 @@ const loadVillageField = () => {
         false
     ));
 
+    GAME.addEntity(new FoodDrop(
```
Comment (calebkrauter) [inline js/scripts/zone_loaders/village_zone_loaders.js:144]: Nice. Do we want this in the village though?

Comment (calebkrauter) [reply js/scripts/zone_loaders/village_zone_loaders.js:144]: You'll be able to farm it too easily I'm thinking. What if the eel boss drops it or you hide it on the map in a place that's hard to get to.

File: js/scripts/zone_loaders/village_zone_loaders.js [hunk 21]
```diff
@@ -141,6 +141,12 @@ const loadVillageField = () => {
...
+        Vector.blockToWorldSpace(new Vector(5, aboveGroundLevel + 4), -1), 
+        FoodDrop.ROAST_TURKEY, 
+        false
+    ));
+
     setTimeout(() => {
```
Comment (calebkrauter) [inline js/scripts/zone_loaders/village_zone_loaders.js:150]: Do we really need set timeout here?

File: js/scripts/zone_loaders/village_zone_loaders.js [hunk 22]
```diff
@@ -319,13 +325,27 @@ const loadVillageMain = () => {
...
-        GAME.addEntity(new RuneDrop(Vector.blockToWorldSpace(new Vector(66, aboveGroundLevel - 4)), RuneDrop.GRAY, false));
+        GAME.addEntity(new RuneDrop(Vector.blockToWorldSpace(new Vector(46, aboveGroundLevel - 4)), RuneDrop.GREEN, false));
+        GAME.addEntity(new RuneDrop(Vector.blockToWorldSpace(new Vector(73, aboveGroundLevel - 4)), RuneDrop.GRAY, false));
 
         GAME.addEntity(new RuneDrop(Vector.blockToWorldSpace(new Vector(93, aboveGroundLevel - 4)), RuneDrop.WHITE, false));
         GAME.addEntity(new RuneDrop(Vector.blockToWorldSpace(new Vector(95, aboveGroundLevel - 3)), RuneDrop.WHITE, false));
         GAME.addEntity(new RuneDrop(Vector.blockToWorldSpace(new Vector(97, aboveGroundLevel - 2)), RuneDrop.WHITE, false));
 
+        GAME.addEntity(new OculiBot(Vector.blockToWorldSpace(
+            new Vector(65, aboveGroundLevel - 4)),
```
Comment (calebkrauter) [inline js/scripts/zone_loaders/village_zone_loaders.js:336]: Is it temporary that these are in the village?

**Review by calebkrauter — APPROVED** [review]
Looks good! 

Major notes.
- Can we make the turkey hard to get multiple times since it increases max health.
- Can we not spawn bots in the village?
